File: Backend/internal/observability/sentry_test.go
package observability

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"testing"
	"unicode/utf8"

	"github.com/getsentry/sentry-go"
)

func TestInitSentry_NoDSN(t *testing.T) {
	t.Setenv("SENTRY_DSN", "")
	flush, enabled := InitSentry()
	if enabled {
		t.Fatal("DSN 未設定なのに enabled=true")
	}
	flush() // panicしないこと
}

func TestScrubEvent_RemovesSecrets(t *testing.T) {
	ev := &sentry.Event{
		Request: &sentry.Request{
			Cookies: "session=abc",
			Data:    `{"resume_text":"秘密の職務経歴"}`,
			Headers: map[string]string{
				"Authorization":        "Bearer secret",
				"Cookie":               "a=1",
				"X-Admin-Token":        "admin-secret",
				"X-User-Token":         "user-secret",
				"X-Company-User-Token": "company-secret",
				"X-Internal-Token":     "rag-secret",
				// CloudFront は aliases の `*.shukatsu-ai.jp` 経由で api ドメインにも
				// このヘッダーを付けて到達する。Sentry に残ると公開ALBへの frontend
				// 直叩きで詐称 X-Forwarded-For を署名させられる(#1407)。
				"x-origin-token": "cf-secret",
				"Referer":        "https://shukatsu-ai.jp/verify-email?token=ONETIME",
				"X-Request-ID":   "req-keep",
				"Content-Type":   "application/json",
			},
			QueryString: "email=student@example.com",
		},
	}

	out := scrubEvent(ev, nil)
	if out.Request.Cookies != "" {
		t.Errorf("Cookies = %q, want empty", out.Request.Cookies)
	}
	if out.Request.Data != "" {
		t.Fatalf("Data が残っている: %q", out.Request.Data)
	}
	if out.Request.QueryString != "" {
		t.Fatalf("QueryString が残っている: %q", out.Request.QueryString)
	}
	for _, key := range []string{
		"Authorization", "Cookie", "X-Admin-Token", "X-User-Token",
		"X-Company-User-Token", "X-Internal-Token", "x-origin-token", "Referer",
	} {
		if _, ok := out.Request.Headers[key]; ok {
			t.Errorf("%s が残っている", key)
		}
	}
	if got := out.Request.Headers["X-Request-ID"]; got != "req-keep" {
		t.Fatalf("X-Request-ID = %q, want req-keep", got)
	}
}

func TestScrubEvent_NilSafe(t *testing.T) {
	if scrubEvent(nil, nil) != nil {
		t.Fatal("nil event は nil を返すべき")
	}
}

// 例外メッセージには外部レスポンス本文や AI 出力がそのまま入る経路がある
// （resume_review.go の "rag review failed: %s" など）。Request のスクラブだけでは
// 個人情報の混入を防げない。
func TestScrubEvent_例外メッセージを切り詰める(t *testing.T) {
	long := strings.Repeat("あ", maxSentryMessageRunes+50)
	event := &sentry.Event{
		Message:   long,
		Exception: []sentry.Exception{{Value: "rag review failed: " + long}},
	}

	out := scrubEvent(event, nil)

	if utf8.RuneCountInString(out.Exception[0].Value) > maxSentryMessageRunes+len("…(truncated)") {
		t.Errorf("例外メッセージが切り詰められていない: %d文字", utf8.RuneCountInString(out.Exception[0].Value))
	}
	if !strings.HasSuffix(out.Exception[0].Value, "…(truncated)") {
		t.Error("切り詰めたことが分かる印が無い")
	}
	if utf8.RuneCountInString(out.Message) > maxSentryMessageRunes+len("…(truncated)") {
		t.Error("Message が切り詰められていない")
	}
}

func TestScrubEvent_短いメッセージはそのまま(t *testing.T) {
	event := &sentry.Event{Exception: []sentry.Exception{{Value: "db connection failed"}}}
	if got := scrubEvent(event, nil).Exception[0].Value; got != "db connection failed" {
		t.Errorf("短いメッセージまで加工している: %q", got)
	}
}

// 秘匿ヘッダーの一覧は frontend(lib/sentry.ts)と Backend の2箇所にある。
// 片方へ足して他方を忘れると、同じヘッダーが一方の Sentry にだけ残る
// （X-Origin-Token で実際に起きた）。両者を突き合わせて食い違いで落とす。
func TestSensitiveHeaders_フロントと同期(t *testing.T) {
	src, err := os.ReadFile(filepath.Join("..", "..", "..", "frontend", "lib", "sentry.ts"))
	if err != nil {
		t.Fatalf("frontend/lib/sentry.ts を読めない（移動したならこのテストも直す）: %v", err)
	}
	block := regexp.MustCompile(`(?s)SENSITIVE_HEADERS = new Set\(\[(.*?)\]\)`).FindSubmatch(src)
	if block == nil {
		t.Fatal("frontend/lib/sentry.ts の SENSITIVE_HEADERS を見つけられない")
	}

	front := map[string]bool{}
	for _, m := range regexp.MustCompile(`'([^']+)'`).FindAllSubmatch(block[1], -1) {
		front[string(m[1])] = true
	}
	if len(front) == 0 {
		t.Fatal("frontend 側のヘッダーを1つも抽出できていない（記法が変わった可能性）")
	}

	for h := range front {
		if !sensitiveHeaders[h] {
			t.Errorf("frontend にだけある: %s（Backend の sensitiveHeaders にも足す）", h)
		}
	}
	for h := range sensitiveHeaders {
		if !front[h] {
			t.Errorf("Backend にだけある: %s（frontend の SENSITIVE_HEADERS にも足す）", h)
		}
	}
}

File: frontend/tests/lib/sentry.test.ts
import { scrubSentryBreadcrumb, scrubSentryEvent } from '@/lib/sentry'
import type { Breadcrumb, ErrorEvent } from '@sentry/nextjs'

describe('scrubSentryEvent', () => {
  it('認証ヘッダーとボディを落とす', () => {
    const event = {
      request: {
        data: { resume_text: '秘密' },
        cookies: { session: 'abc' },
        query_string: 'email=a@b.com',
        headers: {
          Authorization: 'Bearer secret',
          Cookie: 'a=1',
          'X-Admin-Token': 'admin',
          'X-Request-ID': 'keep',
          'Content-Type': 'application/json',
        },
      },
    } as unknown as ErrorEvent

    const out = scrubSentryEvent(event)
    expect(out).not.toBeNull()
    expect(out?.request?.data).toBeUndefined()
    expect(out?.request?.cookies).toBeUndefined()
    expect(out?.request?.query_string).toBe('')
    expect(out?.request?.headers?.Authorization).toBeUndefined()
    expect(out?.request?.headers?.Cookie).toBeUndefined()
    expect(out?.request?.headers?.['X-Admin-Token']).toBeUndefined()
    expect(out?.request?.headers?.['X-Request-ID']).toBe('keep')
  })
})

// httpContextIntegration が window.location.href をそのまま request.url に入れるため、
// query_string を空にするだけでは足りない。クエリにトークンを載せる画面が実在する:
//   /verify-email?token= / /company-portal/setup?token= / /auth/callback?user=
describe('scrubSentryEvent の URL', () => {
  it('request.url のクエリを落とす', () => {
    const event = {
      request: { url: 'https://shukatsu-ai.jp/verify-email?token=ONETIME', query_string: 'token=ONETIME' },
    } as unknown as ErrorEvent

    const out = scrubSentryEvent(event)
    expect(out?.request?.url).toBe('https://shukatsu-ai.jp/verify-email')
    expect(out?.request?.url).not.toContain('ONETIME')
  })

  it('Referer も落とす（遷移元のクエリにトークンが載る）', () => {
    const event = {
      request: { headers: { Referer: 'https://shukatsu-ai.jp/auth/callback?user=BASE64' } },
    } as unknown as ErrorEvent

    expect(scrubSentryEvent(event)?.request?.headers?.Referer).toBeUndefined()
  })
})

// パンくずは beforeSend の対象外。fetch/navigation の URL をそのまま持つ。
describe('scrubSentryBreadcrumb', () => {
  it('fetch のURLからクエリを落とす', () => {
    const out = scrubSentryBreadcrumb({
      category: 'fetch',
      data: { url: '/api/auth/verify-email?token=ONETIME' },
    } as Breadcrumb)
    expect(out?.data?.url).toBe('/api/auth/verify-email')
  })

  it('画面遷移の from/to からクエリを落とす', () => {
    const out = scrubSentryBreadcrumb({
      category: 'navigation',
      data: { from: '/results?session=abc', to: '/verify-email?token=ONETIME' },
    } as Breadcrumb)
    expect(out?.data?.from).toBe('/results')
    expect(out?.data?.to).toBe('/verify-email')
  })

  it('console のパンくずは丸ごと捨てる（何が載るか読めない）', () => {
    expect(scrubSentryBreadcrumb({ category: 'console', message: '秘密' } as Breadcrumb)).toBeNull()
  })
})
